# Spam Detector

In [91]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import numpy as np

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [92]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*My Prediction - Random Forests Classifer

**When to Use Which**
**Use Logistic Regression if:
- Your data has a linear relationshipw with the outcome
- You need a model that is easy to interpret
- Computational efficiency is a concern and you want a model that's fast to train and predict
**Use Random Forests if:
- Your data is complex with non-linear relationships and interactions
- You prioritize predictive performance over interpretability
- You have enough computational resources and are dealing with large datasets
**In practice, it's often useful to start with a simpler model like logistic regression and then move to more complex models like random forests if the simpler model doesn't perform well.

## Split the Data into Training and Testing Sets

In [93]:
# Create the labels set `y` and features DataFrame `X`
X = data.copy()
X = X.drop(columns="spam")
y = data["spam"]

In [94]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()


spam
0    2788
1    1813
Name: count, dtype: int64

In [95]:
print(data.dtypes)

word_freq_make                float64
word_freq_address             float64
word_freq_all                 float64
word_freq_3d                  float64
word_freq_our                 float64
word_freq_over                float64
word_freq_remove              float64
word_freq_internet            float64
word_freq_order               float64
word_freq_mail                float64
word_freq_receive             float64
word_freq_will                float64
word_freq_people              float64
word_freq_report              float64
word_freq_addresses           float64
word_freq_free                float64
word_freq_business            float64
word_freq_email               float64
word_freq_you                 float64
word_freq_credit              float64
word_freq_your                float64
word_freq_font                float64
word_freq_000                 float64
word_freq_money               float64
word_freq_hp                  float64
word_freq_hpl                 float64
word_freq_ge

In [96]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X,y)
X_train.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
956             0.00               0.00           0.00           0.0   
4482            0.00               0.11           0.11           0.0   
3699            0.00               0.00           0.00           0.0   
415             0.62               0.00           0.62           0.0   
2514            0.00               0.00           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
956            3.09             0.0              1.03                1.03   
4482           0.34             0.0              0.00                0.00   
3699           0.53             0.0              0.00                0.00   
415            1.25             0.0              0.62                0.00   
2514           0.00             0.0              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
956              0.00            1.03  ...                   0.0        0.000   
4482             0.11            0.69  ...                   0.0        0.017   
3699             0.00            0.00  ...                   0.0        0.000   
415              0.00            0.00  ...                   0.0        0.000   
2514             0.00            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
956         0.000          0.0        0.000        0.000        0.000   
4482        0.153          0.0        0.017        0.068        0.017   
3699        0.188          0.0        0.000        0.000        0.000   
415         0.000          0.0        0.456        0.273        0.000   
2514        0.502          0.0        0.000        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
956                       10.692                          65   
4482                       3.441                          35   
3699                       1.142                           3   
415                        2.520                          22   
2514                       1.000                           1   

      capital_run_length_total  
956                        139  
4482                       499  
3699                        40  
415                        121  
2514                         8  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [97]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()


In [98]:
# Fit the Standard Scaler with the training data
scaler = StandardScaler().fit(X_train)

In [99]:
# Scale the training data
X_train_scaled = scaler.transform(X_train)
X_train_scaled

X_test_scaled = scaler.transform(X_test)
X_test_scaled

array([[-0.33589697, -0.17013957, -0.55572181, ..., -0.12236178,
        -0.24406594, -0.47389147],
       [-0.33589697, -0.17013957, -0.55572181, ..., -0.07807808,
        -0.19726245, -0.41727221],
       [-0.33589697,  0.10227139,  0.83825272, ..., -0.08856404,
        -0.2113035 , -0.32976973],
       ...,
       [-0.33589697, -0.17013957, -0.55572181, ...,  0.02880366,
         0.21460823,  0.61903171],
       [-0.33589697, -0.17013957, -0.55572181, ..., -0.12236178,
        -0.24406594, -0.47389147],
       [-0.33589697, -0.17013957, -0.55572181, ...,  0.00806283,
        -0.08961443, -0.17020638]])

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [100]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

logistic_regression_model = LogisticRegression(random_state=1)
logistic_regression_model




LogisticRegression(random_state=1)

In [101]:
logistic_regression_model = logistic_regression_model.fit(X_train_scaled, y_train)


In [102]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = logistic_regression_model.predict(X_test_scaled)

# Review the predictions
testing_predictions


array([0, 0, 1, ..., 1, 0, 0], dtype=int64)

In [104]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predictions)


0.9278887923544744

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [105]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
random_forest_model = RandomForestClassifier(random_state=1)
random_forest_model


RandomForestClassifier(random_state=1)

In [106]:
random_forest_model = random_forest_model.fit(X_train_scaled, y_train)

In [107]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = random_forest_model.predict(X_test_scaled)

# Review the predictions
testing_predictions


array([0, 0, 1, ..., 1, 0, 0], dtype=int64)

In [108]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predictions)


0.9496090356211989

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

**The Random Forest Classifier Model performed higher with an accuracy score of 0.9496090356211989 vs. an accuracy score of 0.9278887923544744 for the Logistic Regression Model.  
**The higher percentage indicates that the Random Forest Classifer Model is more relianble for predicting whether or not an e-mail is spam.